tests/test_metadata_tools.py

```diff
@@ -1,27 +1,92 @@
+from EXIF_constants import EXIFConstants
+import xml.etree.ElementTree as ET
+import subprocess
+
 class TestMetadataTools(unittest.TestCase):
 
+    @classmethod
+    def get_exif_tags(self):
+        """Helper method to extract complete list of exif terms compatible with exiftool"""
+        result = subprocess.run(['exiftool', '-listx', '-f'], capture_output=True, text=True)
+
+        if result.returncode != 0:
+            raise Exception(f"Error running exiftool: {result.stderr}")
+
+        xml_output = result.stdout
+
+        #XML output --> string
+        root = ET.fromstring(xml_output)
+
+        tags = set()
+        for table in root.findall('table'):
+            for tag in table.findall('tag'):
+                tag_name = tag.get('name')
+                tags.add(tag_name)
+
+            # Check for IFD0 group
+            if table.get('name') == 'IFD0':
+                for tag in table.findall('tag'):
+                    tag_name = f"IFD0:{tag.get('name')}"
+                    tags.add(tag_name)
+
+        return tags
+
+
     def setUp(self):
         self.path = "tests/test_images/test_image.jpg"
         self.md = MetadataTools(path=self.path)
+        self.valid_exif_tags = self.get_exif_tags()
         shutil.copyfile("tests/test_images/test_image.jpg", "tests/test_images/image_backup.jpg")
 
+
     def test_read_exif_tags(self):
         """tests exif read function"""
         exif_dict = self.md.read_exif_tags()
         self.assertFalse(pd.isna(exif_dict))
         self.assertNotEqual(exif_dict, {})
-        self.assertEqual(exif_dict[272], 'iPhone XR')
+        self.assertEqual(exif_dict['EXIF:LensMake'], 'Apple')
 
     def test_write_exif_tags(self):
-        pass
+        """tests exif attach function"""
+        datetime_test = datetime.now()
+        datetime_test = datetime_test.strftime('%Y:%m:%d %H:%M:%S.%f')
+        exif_dict = {"XMP:CreatorCity": 'San Francisco', 'IPTC:CopyrightNotice': 'CaliforniaAcademy',
+                     'XMP:CreateDate': f'{datetime_test}'}
+        self.md.write_exif_tags(exif_dict=exif_dict)
+        exif_return = self.md.read_exif_tags()
+        print(exif_return)
+        self.assertEqual("San Francisco", exif_return['XMP:CreatorCity'])
+        self.assertEqual('CaliforniaAcademy', exif_return['IPTC:CopyrightNotice'])
+        self.assertEqual(str(f'{datetime_test}'), str(exif_return['XMP:CreateDate']))
+
+    def test_invalid_exif_tags(self):
+        exif_dict = {'EXIF:TEST1': 'Samsung', 'IPTC:CopyrightNotice': 'CaliforniaAcademy', 'EXIF:ApertureValid': '1.5'}
+        with self.assertRaises(ValueError) as context:
+            self.md.write_exif_tags(exif_dict=exif_dict)
+
+        self.assertEqual(str(context.exception),
+                         "Invalid keys in exif_dict, check exif "
+                         "constants:{'EXIF:TEST1': False, 'IPTC:CopyrightNotice': True, 'EXIF:ApertureValid': False}")
+
 
     def test_accepted_exif_tags(self):
```

Comment:
tests whether write exif will reject inccorrect exif tags and return correct error.